### PR TITLE
Refactor metadata emission to avoid visiting HIR

### PR DESCRIPTION
**Pull request by cjgillot**
This PR refactors metadata emission to be based on tables and iteration over definitions.

In a first part, this PR moves information from the `EntryKind` enum to tables, until removing the `EntryKind` enum.
In a second part, the iteration scheme is refactored to avoid fetching HIR unless strictly necessary.

r? @ghost

### CODE REVIEW

**Comment by wesleywiser** [comment]
Hi @cjgillot! This came up in the compiler team triage meeting since it has been sitting a while without reviews. It looks like you're still iterating on this so whenever you're ready for review, just roll for a reviewer. 🙂

**Comment by oli-obk** [comment]
From the diff it looks to me like we should avoid running metadata encoding if none of its dependencies were dirty/red

If that is not feasible for now, I think we should still land it, as the regressions are limited to incremental cases and the code is cleaner now

File: compiler/rustc_middle/src/query/mod.rs [hunk 1]
```diff
@@ -1229,6 +1229,7 @@ rustc_queries! {
         desc { |tcx| "looking up function parameter names for `{}`", tcx.def_path_str(def_id) }
         separate_provide_extern
     }
+
```
Comment (oli-obk) [inline compiler/rustc_middle/src/query/mod.rs:1232]: This commit does nothing

**Review by oli-obk — APPROVED** [review]
there's a stray empty commit.

The perf results look mixed, but not negative anymore, so let's get this landed!

**Comment by Kobzol** [comment]
[Metadata](https://perf.rust-lang.org/compare.html?start=eb76764ea412b36f670564f5d58c317cec8fa496&end=5bd28f5eac1ba3569bfa8d49ec3f5acbdfdff7a0&stat=size%3Acrate_metadata&tab=compile) and [binary size](https://perf.rust-lang.org/compare.html?start=eb76764ea412b36f670564f5d58c317cec8fa496&end=5bd28f5eac1ba3569bfa8d49ec3f5acbdfdff7a0&stat=size%3Alinked_artifact&tab=compile) has regressed somewhat, which was probably expected given this change (?). Max RSS [improvements](https://perf.rust-lang.org/compare.html?start=eb76764ea412b36f670564f5d58c317cec8fa496&end=5bd28f5eac1ba3569bfa8d49ec3f5acbdfdff7a0&stat=max-rss&tab=compile) look really nice!

**Comment by oli-obk** [comment]
Hmm, we can probably avoid the metadata regression by checking which items we're unnecessarily encoding now.

**Pull request merged**